pkg/types/portmap.go

```diff
@@ -47,21 +49,75 @@ func (pps PortProtoSet) Equal(other PortProtoSet) bool {
 	return true
 }
 
+// Add increments the reference count for the specified key.
+func (pps PortProtoSet) Add(pp PortProto) bool {
+	return counter.Counter[PortProto](pps).Add(pp)
+}
+
+// Delete decrements the reference count for the specified key.
+func (pps PortProtoSet) Delete(pp PortProto) bool {
+	return counter.Counter[PortProto](pps).Delete(pp)
+}
+
 // NamedPortMultiMap may have multiple entries for a name if multiple PODs
 // define the same name with different values.
-type NamedPortMultiMap map[string]PortProtoSet
+type NamedPortMultiMap interface {
+	// GetNamedPort returns the port number for the named port, if any.
+	GetNamedPort(name string, proto uint8) (uint16, error)
+	// Len returns the number of Name->PortProtoSet mappings known.
+	Len() int
+}
 
-// Equal returns true if the NamedPortMultiMaps are equal.
-func (npm NamedPortMultiMap) Equal(other NamedPortMultiMap) bool {
-	if len(npm) != len(other) {
-		return false
+// NamedPortMultiMapUpdater allows for mutation of the NamedPortMultiMap, and is
+// intended for ipcache internal usage.
+type NamedPortMultiMapUpdater interface {
+	NamedPortMultiMap
+	Update(old, new NamedPortMap) (namedPortChanged bool)
+}
+
+// Implements both NamedPortMultiMap and its Updater variant. All accesses to it
+// must be protected by its RW mutex.
+type namedPortMultiMap struct {
+	lock.RWMutex
+	m map[string]PortProtoSet
+}
+
+func NewNamedPortMultiMapUpdater() NamedPortMultiMapUpdater {
+	return &namedPortMultiMap{
+		m: make(map[string]PortProtoSet),
 	}
-	for name, ports := range npm {
-		if otherPorts, exists := other[name]; !exists || !ports.Equal(otherPorts) {
-			return false
+}
+
+func (npm *namedPortMultiMap) Len() int {
+	npm.RLock()
+	defer npm.RUnlock()
+	return len(npm.m)
+}
+
+// Update applies potential changes in named ports, and returns whether there were any.
+func (npm *namedPortMultiMap) Update(old, new NamedPortMap) (namedPortsChanged bool) {
+	npm.Lock()
+	defer npm.Unlock()
+	// The order is important here. Increment the refcount first, and then
+	// decrement it again for old ports, so that we don't hit zero if there are
+	// no changes.
+	for name, port := range new {
+		c, ok := npm.m[name]
+		if !ok {
+			c = make(PortProtoSet)
+			npm.m[name] = c
+		}
+		if c.Add(port) {
+			namedPortsChanged = true
 		}
 	}
-	return true
+	for name, port := range old {
+		if PortProtoSet(npm.m[name]).Delete(port) {
+			delete(npm.m, name)
```

Comment:
I don't think this is right. Even if we deleted a port, there might be other (non-conflicting) ports with the same name. We only should do a `delete(npm.m, name)` if `len(npm.m) == 0`.

Here's a unit test to demonstrate what I mean:

It fails at the end, because `http=8080:UDP` is no longer present, even though it has not been deleted.
```go

func (ds *PortsTestSuite) TestPolicyNamedPortMultiMapUpdate(c *C) {
	npm := NewNamedPortMultiMapUpdater()

	pod1PortsOld := map[string]PortProto{}
	pod1PortsNew := map[string]PortProto{
		"http": {80, uint8(u8proto.TCP)},
	}

	// Insert http=80:TCP from pod1
	changed := npm.Update(pod1PortsOld, pod1PortsNew)
	c.Assert(changed, Equals, true)

	// No changes
	changed = npm.Update(pod1PortsNew, pod1PortsNew)
	c.Assert(changed, Equals, false)

	// Assert http=80:TCP
	port, err := npm.GetNamedPort("http", uint8(u8proto.TCP))
	c.Assert(err, IsNil)
	c.Assert(port, Equals, uint16(80))

	pod2PortsOld := map[string]PortProto{}
	pod2PortsNew := map[string]PortProto{
		"http": {8080, uint8(u8proto.UDP)},
	}

	// Insert http=8080:UDP from pod2, retain http=80:TCP from pod1
	changed = npm.Update(pod2PortsOld, pod2PortsNew)
	c.Assert(changed, Equals, true)

	port, err = npm.GetNamedPort("http", uint8(u8proto.TCP))
	c.Assert(err, IsNil)
	c.Assert(port, Equals, uint16(80))
	port, err = npm.GetNamedPort("http", uint8(u8proto.UDP))
	c.Assert(err, IsNil)
	c.Assert(port, Equals, uint16(8080))

	// Delete http=80:TCP from pod1, retain http=8080:UDP from pod2
	pod1PortsOld = pod1PortsNew
	pod1PortsNew = map[string]PortProto{}

	// Delete http=80:TCP from pod1
	changed = npm.Update(pod1PortsOld, pod1PortsNew)
	c.Assert(changed, Equals, true)

	port, err = npm.GetNamedPort("http", uint8(u8proto.UDP))
	c.Assert(err, IsNil)
	c.Assert(port, Equals, uint16(8080))
}
```

Reply:
Nice catch. Added the testcase, fixed the problem - PTAL.